### ISSUE TITLE
Automatically fallback in non-TTY environments

### ISSUE THREAD

**Issue by SamVerschueren** [SamVerschueren/listr#33]
This has been playing in my head for quite a while now. So we have support for custom (3rd party) renderers like the current embedded [update-renderer](https://github.com/SamVerschueren/listr-update-renderer) and the [verbose-renderer](https://github.com/SamVerschueren/listr-verbose-renderer). The default renderer is the update renderer, also in a non-TTY environment which doesn't work well with ANSI escape codes.

This means that you will have to write something like this.

``` js
new Listr([
    // tasks go here
], {
    renderer: process.isTTY ? 'update': 'verbose'
});
```

I want to get rid of this and automatically fall back to the `verbose` renderer if Listr internally detects that it is not running in a TTY. This is perfectly possible and I believe that we should really do this. It doesn't make sense to render fancy stuff with `log-update` for instance of it doesn't work correctly.

But what should be done with custom renderers then? Let's first recap what a custom renderer is and how it can be used. If you look into the [custom renderers](https://github.com/SamVerschueren/listr#custom-renderers) section of the readme, you can see that the basic structure looks like this.

``` js
class CustomRenderer {

    constructor(tasks, options) { }

    render() { }

    end(err) { }
}
```

We can then use it and pass it in as `renderer`.

``` js
const custom = require('custom-renderer');

new Listr([
    // tasks go here
], {
    renderer: custom
});
```

What should we do now when executed in a non-TTY environment?
### Option 1. Always use the `verbose` renderer

We could always use the `verbose` renderer by default. The drawback is that you will have to live with it that the output always looks the same.
### Option 2. Custom renderer lets us know that it supports those environments

The custom renderer could set a property indicating if it supports non-TTY environments.

``` js
class CustomRenderer {

    constructor(tasks, options) { }

    get nonTTY() {
        return true;
    }

    render() { }

    end(err) { }
}
```

If Listr detects that it isn't running in a TTY, it will look check if `nonTTY` returns `true`, if that is the case it could safely use that renderer. If it returns a falsy value, it will fall back to the verbose renderer.
### Option 3. Fallback renderer

If you pass in a custom renderer, it will automatically use that renderer in all environments. Except if you pass in another renderer as fallback renderer.

``` js
new Listr([
    // tasks go here
], {
    renderer: custom,
    fallbackRenderer: 'verbose'
});
```

Options to be decided. Could as well be something like

``` js
new Listr([
    // tasks go here
], {
    renderer: {
        default: custom,
        noTTY: 'verbose'
    }
});
```

Any feedback or input on these ideas would be more then appreciated!

// @sindresorhus @jfmengels @kevva @andywer @okonet

**Comment by andywer** [comment]
My first idea was to let the renderer check if it is run in a TTY or not and then switch to another renderer if necessary. But that would be a lot of redundant code in the renderers and error-prone (if forgot to check), of course.

👉 So my favorite option would be 2. Option 3 moves too much implementation details into the hands of the user for my taste.

Let the user select a custom renderer and let then the custom renderer decide what to do (if it supports non-TTY). If it does state support for non-TTY then fall back to verbose renderer.

Alternative: You could also let the custom renderers extend a base class that cares for the non-TTY fallback, but to be honest, I am not a big fan of class-based inheritance.

**Comment by SamVerschueren** [comment]
Thanks for the feedback @andywer.

Another option that actually extends option number 2 is that the `nonTTY` property could also return the renderer that should be used in non-TTY environments. If it is not implemented, it will fallback to `verbose`, otherwise it will use the one returned by the property.

```js
const fooRenderer = require('foo-renderer');

class CustomRenderer {

    constructor(tasks, options) { }

    get nonTTY() {
        return fooRenderer;
    }

    render() { }

    end(err) { }
}
```

**Comment by andywer** [comment]
Yeah, but then you should better make sure you cannot run into an infinite loop.

It might be easier to just let `nonTTY` be true and then let the renderer code do the fallback.

**Comment by okonet** [comment]
I would go for 3 since it's the simplest and most extensible one. In the option 2 I see the lack of extensibility, especially when the renderer returns another renderer. I can imaging how the next request after this one is implemented will be "allow channging the fallback renderer" so you end up with both options implemented :)

IMO verbose renderer will fit 99% users. For the rest a simple option (`ttyRenderer` is more clear BTW) will be enough.

**Comment by sindresorhus** [comment]
I prefer `2`. The custom renderer is the most knowledgeable whether it works in non-TTY or not, not the user.

**Comment by SamVerschueren** [comment]
Maybe option 2 with the possibility to change the non-TTY renderer. If that fallback renderer also returns `false`, it will just go for the default verbose renderer.

**Comment by SamVerschueren** [comment]
Thank you all for the feedback. I'm going to add option number `2` with a boolean because as @sindresorhus noted, renderers have the knowledge of being able to run in non-TTY, not users/developers.

The only thing I have to decide is how to define fallback renderers. Might go with an array.

```js
new Listr([
    // tasks go here
], {
    renderer: custom
});
```

And if you need a fallback

```js
new Listr([
    // tasks go here
], {
    renderer: [custom, verbose]
});
```

If you want multiple fallbacks, in case verbose returns `false` for non-TTY environments as well, you can.

```js
new Listr([
    // tasks go here
], {
    renderer: [custom, verbose, unicorn]
});
```

It would work the same way as fonts in CSS. If none of the renderers return `true`, it will pick the built-in `verbose` one.

**Comment by okonet** [comment]
Will it also fallback by default or renderers have to be updated to gain the new functionality?

**Comment by SamVerschueren** [comment]
A renderer has to implement the `get nonTTY`. If it does not, it's the same as returning a falsy value and it will automatically fallback to the "next" one.

**Comment by sindresorhus** [comment]
> If you want multiple fallbacks, in case verbose returns false for non-TTY environments as well, you can.

So when would `verbose` be used then? If `custom` is used in TTY, and both `custom` and `verbose` returns `false` for non-TTY.

**Comment by sindresorhus** [comment]
I still don't get it. There are only two states, `nonTTY === true` and `nonTTY === false`. What's the point of specifying more than 1 TTY renderer?

**Comment by SamVerschueren** [comment]
You are right, there is no point in doing that. And probably most of the time people would have a TTY and a non-TTY renderer specified. Or if they don't care, only a custom TTY renderer. But maybe the `unicorn` renderer receives an update and suddenly doesn't support non-TTY mode anymore.

The thing is that I was just explaining how it would work. I don't say it's a good thing of specifying 10 renderers, actually quite the opposite because 9 of them will never be used.

It was a suggesting how things could be implemented. I'm still undecided about it. Would you be more in favour of adding a separate option for nonTTY?

```js
new Listr([
    // tasks go here
], {
    renderer: custom,
    nonTTYRenderer: unicorn
});
```

**Comment by sindresorhus** [comment]
> Would you be more in favour of adding a separate option for nonTTY?

Yes, I think that would be clearer.

**Comment by okonet** [comment]
I also think this behavior implies too much magic. So if something isn't working on my CI it's hard to say why. If there is `nonTTYRenderer` option (or similar, I'm bad at naming), it becomes clear. I still think it should fallback to `verbose` automatically if this isn't set since I believe 99% of users won't care and expect things to just work.

**Comment by SamVerschueren** [comment]
> I still think it should fallback to verbose automatically

It will :)

Thanks for the feedback everyone! Will go for a separate option to make it clear. I will go for `nonTTYRenderer` for now unless someone comes up with something better :).

**Comment by SamVerschueren** [comment]
@okonet Just to be clear. It will if the default renderer does not support `nonTTY`. If it supports non-TTY environments, it will use that one. If it doesn't, it will automatically fallback to `verbose`.

**Comment by okonet** [comment]
@SamVerschueren awesome news! Can you publish a new version please? I'd really like this to be a part of https://github.com/okonet/lint-staged/pull/108

**Comment by SamVerschueren** [comment]
@okonet could you wait a little longer? Want to fix #34 first as well. Hope I can start it this evening.

**Comment by okonet** [comment]
That would be even better although I'm all for smaller but more recent releases :)

**Comment by SamVerschueren** [comment]
@okonet You're right. If I'm not able to do it this evening, I will release it instead. Thanks for the feedback!

**Comment by SamVerschueren** [comment]
@okonet released in `v0.9.0`.

**Comment by okonet** [comment]
Awesome! Thank you!!!
